lib/block/merkleblock.js

```diff
@@ -0,0 +1,284 @@
+  $.checkState(_.isArray(this.hashes), 'MerkleBlock hashes is not an array');
+  if(_.isBoolean(this._validMerkleTree)) {
+    return this._validMerkleTree;
+  }
+
+  // Can't have more hashes than numTransactions
+  if(this.hashes.length > this.numTransactions) {
+    return this._setValidMerkleTree(false);
+  }
+
+  // Can't have more flag bits than num hashes
+  if(this.flags.length * 8 < this.hashes.length) {
+    return this._setValidMerkleTree(false);
+  }
+
+  var height = 0;
+  while (this._calcTreeWidth(height) > 1) {
+    height++;
+  }
+
+  this._flagBitsUsed = 0;
+  this._hashesUsed = 0;
+  var root = this._traverseMerkleTree(height, 0);
+  if(this._hashesUsed !== this.hashes.length) {
+    return this._setValidMerkleTree(false);
+  }
+  return this._setValidMerkleTree(BufferUtil.equals(root, this.header.merkleRoot));
+}
+
+/** Traverse a the tree in this MerkleBlock, validating it along the way
+ *  Modeled after Bitcoin Core merkleblock.cpp TraverseAndExtract()
+ * @param {Number} - Current height
+ * @param {Number} - Current position in the tree
+ * @returns {Buffer|null} - Buffer containing the Merkle Hash far that height
+ * @private
+ */
+MerkleBlock.prototype._traverseMerkleTree = function traverseMerkleTree(depth, pos) {
+  if(this._flagBitsUsed > this.flags.length * 8) {
+    return null;
+  }
+  var isParentOfMatch = (this.flags[this._flagBitsUsed >> 3] >>> (this._flagBitsUsed++ & 7)) & 1;
+  if(depth === 0 || !isParentOfMatch) {
+    if(this._hashesUsed >= this.hashes.length) {
+      return null;
+    }
+    var hash = this.hashes[this._hashesUsed++];
+    return new Buffer(hash, 'hex');
+  } else {
+    var left = this._traverseMerkleTree(depth-1, pos*2);
+    var right;
+    if(pos*2+1 < this._calcTreeWidth(depth-1)) {
+      right = this._traverseMerkleTree(depth-1, pos*2+1);
+    } else {
+      right = left;
+    }
+    return Hash.sha256sha256(new Buffer.concat([left, right]));
+  }
+}
+
+/** Calculates the width of a merkle tree at a given height.
+ *  Modeled after Bitcoin Core merkleblock.h CalcTreeWidth()
+ * @param {Number} - Height at which we want the tree width
+ * @returns {Number} - Width of the tree at a given height
+ * @private
+ */
+MerkleBlock.prototype._calcTreeWidth = function calcTreeWidth(height) {
+  return (this.numTransactions + (1 << height) - 1) >> height;
+}
+
+/**
+ * @param {Transaction|String} - Transaction or Transaction ID Hash
+ * @returns {Bool} - return true/false if this MerkleBlock has the TX or not
+ * @private
+ */
+MerkleBlock.prototype.hasTransaction = function hasTransaction(tx) {
+  $.checkArgument(!_.isUndefined(tx), 'tx cannot be undefined');
+  $.checkArgument(tx instanceof Transaction || typeof tx === 'string',
+      'Invalid tx given, tx must be a "string" or "Transaction"');
+
+  var hash = tx.id || tx;
+  var revHash = BufferUtil.reverse(new Buffer(hash,'hex')).toString('hex');
+
+  return (this.hashes.indexOf(hash) !== -1
```

Comment:
So it would be more similar to:

```javascript
var hashes = this.getTransactionHashes();
if (hashes) {
  return (hashes.indexOf(hash) !== -1);
}
return false;
```